pkg/gpu/consumer.go

```diff
@@ -0,0 +1,167 @@
+	"github.com/DataDog/datadog-agent/pkg/process/monitor"
+	"github.com/DataDog/datadog-agent/pkg/status/health"
+	"github.com/DataDog/datadog-agent/pkg/util/kernel"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// cudaEventConsumer is responsible for consuming CUDA events from the eBPF probe, and delivering them
+// to the appropriate stream handler.
+type cudaEventConsumer struct {
+	eventHandler                    ddebpf.EventHandler
+	once                            sync.Once
+	closed                          chan struct{}
+	streamHandlers                  map[model.StreamKey]*StreamHandler
+	wg                              sync.WaitGroup
+	scanTerminatedProcessesInterval time.Duration
+}
+
+// NewCudaEventConsumer creates a new CUDA event consumer.
+func NewCudaEventConsumer(eventHandler ddebpf.EventHandler, scanTerminatedProcessesInterval time.Duration) *cudaEventConsumer {
+	return &cudaEventConsumer{
+		eventHandler:                    eventHandler,
+		closed:                          make(chan struct{}),
+		streamHandlers:                  make(map[model.StreamKey]*StreamHandler),
+		scanTerminatedProcessesInterval: scanTerminatedProcessesInterval,
+	}
+}
+
+// Stop stops the CUDA event consumer.
+func (c *cudaEventConsumer) Stop() {
+	if c == nil {
+		return
+	}
+	c.eventHandler.Stop()
+	c.wg.Wait()
+	c.once.Do(func() {
+		close(c.closed)
+	})
+}
+
+// Start starts the CUDA event consumer.
+func (c *cudaEventConsumer) Start() {
+	if c == nil {
+		return
+	}
+	health := health.RegisterLiveness("gpu-tracer-cuda-events")
+	processMonitor := monitor.GetProcessMonitor()
+	cleanupExit := processMonitor.SubscribeExit(c.handleProcessExit)
+
+	c.wg.Add(1)
+	go func() {
+		processSync := time.NewTicker(c.scanTerminatedProcessesInterval)
+
+		defer func() {
+			cleanupExit()
+			err := health.Deregister()
+			if err != nil {
+				log.Warnf("error de-registering health check: %s", err)
+			}
+			c.wg.Done()
+			log.Trace("CUDA event consumer stopped")
+		}()
+
+		dataChannel := c.eventHandler.DataChannel()
+		lostChannel := c.eventHandler.LostChannel()
+		for {
+			select {
+			case <-c.closed:
+				return
+			case <-health.C:
+			case <-processSync.C:
+				c.checkClosedProcesses()
+			case batchData, ok := <-dataChannel:
+				if !ok {
+					return
+				}
+
+				if len(batchData.Data) < gpuebpf.SizeofCudaEventHeader {
+					log.Errorf("Not enough data to parse header, data size=%d, expecting at least %d", len(batchData.Data), gpuebpf.SizeofCudaEventHeader)
+					continue
+				}
+
+				header := (*gpuebpf.CudaEventHeader)(unsafe.Pointer(&batchData.Data[0]))
+
+				pid := uint32(header.Pid_tgid >> 32)
+				streamKey := model.StreamKey{Pid: pid, Stream: header.Stream_id}
+
+				if _, ok := c.streamHandlers[streamKey]; !ok {
+					c.streamHandlers[streamKey] = newStreamHandler()
+				}
+
+				switch header.Type {
+				case gpuebpf.CudaEventTypeKernelLaunch:
+					if len(batchData.Data) != gpuebpf.SizeofCudaKernelLaunch {
```

Comment:
you already check the length few lines before, maybe just use a local variable to store it?
also you check for exact match to the size of CudaKernelLaunch, but according to the log message **at least** minimum size is required.

Reply:
same comment for other cases in this switch block

Reply:
Fixed, log message was wrong, it should be exact.